### PR TITLE
Geo/show element isactive in vtk

### LINKED ISSUE #10264: [GeoMechanicsApplication] Add vtk writing of active elements to core
Move the VTK changes from GeoMechanicsBranch to new branch and issue Pull Request. As changes are in core only and are not vital for running of branches.

### PR DESCRIPTION
**Pull request by mcgicjn2**
**📝 Description**
This is a small core change to output if an element is active in VTK files. 

This was originally in our branch but moved and separated to a new branch to allow PR at the core level.

It is a nice feature to have, but is non-critical. 

Change originally made by @Vahid-Galavi (Is this what you intended with this changed ?)

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
- Added "Active" to allow visualization of active elements in VTK

### CODE REVIEW

File: kratos/utilities/entities_utilities.h [hunk 1]
```diff
@@ -118,7 +118,16 @@ namespace EntitiesUtilities
         block_for_each(
             GetEntities<TEntityType>(rModelPart),
             [&r_current_process_info](TEntityType& rEntity){
-                rEntity.InitializeSolutionStep(r_current_process_info);
+                // Detect if the entity is active or not. If the user did not make any choice the entity
```
Comment (loumalouomega) [inline kratos/utilities/entities_utilities.h]: I would separate this changes in an independent PR

Comment (loumalouomega) [reply kratos/utilities/entities_utilities.h]: I mean the entities_utilities

File: kratos/includes/constitutive_law.h [hunk 2]
```diff
@@ -1165,7 +1165,7 @@ class KRATOS_API(KRATOS_CORE) ConstitutiveLaw : public Flags
      * @see Parameters
      * @see StressMeasures
      */
-    void FinalizeMaterialResponse (Parameters& rValues,const StressMeasure& rStressMeasure);
+    virtual void FinalizeMaterialResponse (Parameters& rValues,const StressMeasure& rStressMeasure);
```
Comment (loumalouomega) [inline kratos/includes/constitutive_law.h]: Huuuuuum

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Important changes, blocking to avoid accidental merge

File: kratos/includes/constitutive_law.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: A priori this is already virtual in their different calls:

~~~c++
    /**
     * Finalize the material response in terms of 1st Piola-Kirchhoff stresses
     * @see Parameters
     */

    virtual void FinalizeMaterialResponsePK1 (Parameters& rValues);

    /**
     * Finalize the material response in terms of 2nd Piola-Kirchhoff stresses
     * @see Parameters
     */

    virtual void FinalizeMaterialResponsePK2 (Parameters& rValues);

    /**
     * Finalize the material response in terms of Kirchhoff stresses
     * @see Parameters
     */

    virtual void FinalizeMaterialResponseKirchhoff (Parameters& rValues);

    /**
     * Finalize the material response in terms of Cauchy stresses
     * @see Parameters
     */

    virtual void FinalizeMaterialResponseCauchy (Parameters& rValues);
~~~
Are you adding an additional measure?

Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: Opinion @AlejandroCornejo ?

Comment (AlejandroCornejo) [reply kratos/includes/constitutive_law.h]: well we should check the effect of this elimination of virtual. I think that this method is never overriden but who knows... If it's never overriden could be a good thing this change

Comment (loumalouomega) [reply kratos/includes/constitutive_law.h]: The point is why override, you should override the submethods, except if you have another stress measures, but then maybe you should propose the new stress measure instead of this.

Comment (rubenzorrilla) [reply kratos/includes/constitutive_law.h]: Agree with @loumalouomega observation. This must remain as non-virtual. The ones to be overridden are the corresponding stress measures.

Comment (AlejandroCornejo) [reply kratos/includes/constitutive_law.h]: ah! reviewing it fast I though he that was removing the virtual. If it is adding the virtual I would avoid that unnecessary overhead.

Comment (mcgicjn2) [reply kratos/includes/constitutive_law.h]: Thanks for the quick reply / review I will do the necessary separation tomorrow and remove the virtual.

Comment (mcgicjn2) [reply kratos/includes/constitutive_law.h]: Reverted

File: kratos/utilities/entities_utilities.h (hunk 1 above)
Comment (mcgicjn2) [reply kratos/utilities/entities_utilities.h]: Removed to alternative PR

**Comment by mcgicjn2** [comment]
Hi @loumalouomega,

I have made the necessary changes. Can this be reviewed ? I am having difficulty changing the PR to review required. (Sorry bit of a github novice here) Cheers Jon

File: kratos/includes/gid_gauss_point_container.h [hunk 3]
```diff
@@ -370,100 +491,120 @@ class GidGaussPointsContainer
...
+                                                    ValuesOnIntPoint[index](0,3), ValuesOnIntPoint[index](0,4),
+                                                    ValuesOnIntPoint[index](0,5) );
+                                }
+                                else if(ValuesOnIntPoint[index].size1() ==1
+                                        && ValuesOnIntPoint[index].size2() ==3) {
+                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
+                                                    ValuesOnIntPoint[index](0,1), 0.0,
+                                                    ValuesOnIntPoint[index](0,2), 0.0, 0.0);
+                                }
+                                else if(ValuesOnIntPoint[index].size1() ==1
```
Comment (loumalouomega) [inline kratos/includes/gid_gauss_point_container.h]:
```suggestion
                                } else if(ValuesOnIntPoint[index].size1() ==1
                                        && ValuesOnIntPoint[index].size2() ==6) {
                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
                                                    ValuesOnIntPoint[index](0,1), ValuesOnIntPoint[index](0,2),
                                                    ValuesOnIntPoint[index](0,3), ValuesOnIntPoint[index](0,4),
                                                    ValuesOnIntPoint[index](0,5) );
                                } else if(ValuesOnIntPoint[index].size1() ==1
                                        && ValuesOnIntPoint[index].size2() ==3) {
                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
                                                    ValuesOnIntPoint[index](0,1), 0.0,
                                                    ValuesOnIntPoint[index](0,2), 0.0, 0.0);
                                } else if(ValuesOnIntPoint[index].size1() ==1
```

File: kratos/includes/gid_gauss_point_container.h [hunk 4]
```diff
@@ -370,100 +491,120 @@ class GidGaussPointsContainer
...
+                                        && ValuesOnIntPoint[index].size2() ==4) {
+                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
+                                                    ValuesOnIntPoint[index](0,1), ValuesOnIntPoint[index](0,2),
+                                                    ValuesOnIntPoint[index](0,3), 0.0, 0.0);
+                                }
+                                else if(ValuesOnIntPoint[index].size1() ==1
```
Comment (loumalouomega) [inline kratos/includes/gid_gauss_point_container.h]:
```suggestion
                                } else if(ValuesOnIntPoint[index].size1() ==2
                                        && ValuesOnIntPoint[index].size2() ==2) {
                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
                                                    ValuesOnIntPoint[index](1,1), 0.0,
                                                    ValuesOnIntPoint[index](0,1), 0.0, 0.0);
                                } else if(ValuesOnIntPoint[index].size1() ==1
                                        && ValuesOnIntPoint[index].size2() ==3) {
                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
                                                    ValuesOnIntPoint[index](0,1), 0.0,
                                                    ValuesOnIntPoint[index](0,2), 0.0, 0.0);
                                } else if(ValuesOnIntPoint[index].size1() ==1
                                        && ValuesOnIntPoint[index].size2() ==4) {
                                    GiD_fWrite3DMatrix( ResultFile, it->Id(), ValuesOnIntPoint[index](0,0),
                                                    ValuesOnIntPoint[index](0,1), ValuesOnIntPoint[index](0,2),
                                                    ValuesOnIntPoint[index](0,3), 0.0, 0.0);
                                } else if(ValuesOnIntPoint[index].size1() ==1
```

File: kratos/includes/gid_gauss_point_container.h [hunk 5]
```diff
@@ -370,100 +491,120 @@ class GidGaussPointsContainer
...
+                    } 
                 }
-
-                // Resize first matrix to (0,0) for test below
-                ValuesOnIntPoint[0].resize(0, 0, false);
-
             }
             if( mMeshConditions.size() != 0 ) {
                 for( auto it = mMeshConditions.begin(); it != mMeshConditions.end(); it++ ) {
                     if( !(it->IsDefined(ACTIVE)) || it->Is(ACTIVE) ) {
```
Comment (loumalouomega) [inline kratos/includes/gid_gauss_point_container.h:505]: Once https://github.com/KratosMultiphysics/Kratos/pull/10311 is merged this can be refactored

File: kratos/includes/gid_gauss_point_container.h [hunk 6]
```diff
@@ -323,45 +424,65 @@ class GidGaussPointsContainer
         if( mMeshElements.size() != 0 || mMeshConditions.size() != 0 )
         {
             //WriteGaussPoints(ResultFile);
-            GiD_fBeginResult( ResultFile, (char *)(rVariable.Name()).c_str(), (char *)("Kratos"), SolutionTag,
-                             GiD_Matrix, GiD_OnGaussPoints, mGPTitle, NULL, 0, NULL );
-
-            std::vector<Vector> ValuesOnIntPoint(mSize);
+            bool isEntityActive = false;
```
Comment (loumalouomega) [inline kratos/includes/gid_gauss_point_container.h:427]: Idem, with https://github.com/KratosMultiphysics/Kratos/pull/10311 can be simplified

**Review by loumalouomega — CHANGES_REQUESTED** [review]
This is a behaviour change, and affects many people, I cannot approve alone this. Needs the approval of @KratosMultiphysics/technical-committee. With https://github.com/KratosMultiphysics/Kratos/pull/10311 the code can be simplified

File: kratos/includes/gid_gauss_point_container.h (hunk 5 above)
Comment (mcgicjn2) [reply kratos/includes/gid_gauss_point_container.h:505]: OK @loumalouomega will wait for #10311

File: kratos/includes/gid_gauss_point_container.h (hunk 6 above)
Comment (mcgicjn2) [reply kratos/includes/gid_gauss_point_container.h:427]: OK @loumalouomega will wait for #10311

**Comment by philbucher** [comment]
can you please add a test with active and inactive elements?

**Comment by mcgicjn2** [comment]
As this is a stale branch and the functionality appears not to be required on our side. Perhaps we can close this and related branches I.e.#10309

**Comment by mcgicjn2** [comment]
Closing
